File: cpp/classes.cpp
#include <string>
#include <bits/stdc++.h> 

namespace bug {
    typedef std::string ErrorMessage;
    struct AccountCode {
        std::string _stringValue;
        AccountCode(std::string s) : _stringValue(s) {}
    };
    struct UserName {
        std::string _stringValue;
        UserName(std::string s) : _stringValue(s) {}
    };
 
    std::string validateUserName(UserName userName) {
        std::regex re("^[a-z]+$"); 
        std::smatch match;
        if (!std::regex_match(userName._stringValue, match, re)) {
            return "user name must be l/c: [${userName}]";
        }
        return "true";
    }
    std::string validateAccountCode(AccountCode code) {
        std::regex re("^[AB][0-9]+$"); 
        std::smatch match;
        if (!std::regex_match(code._stringValue, match, re)) {
            return "account code not valid: [${code}]";
        }              
        return "true";
    }
    void validateDetails(UserName userName, AccountCode accountCode) {
        auto msg =  validateAccountCode(accountCode);
        if (msg != "true") throw msg;
        msg = validateUserName(accountCode);
        if (msg != "true") throw msg;
    }
    void test() {
        try {
            const UserName goodName = UserName("bob");
            const AccountCode goodCode =  AccountCode("A1234");
            validateDetails(goodName, goodCode); // what bug makes this throw an exception ???
        }
        catch (std::string ex) {
            std::cout << "Exception " << ex << std::endl;
        }

    }
}

int main(int argc, char** argv) {
    bug::test();
}

File: ts/classes.ts
namespace classes {
    type ErrorMessage = string;
    class AccountCode {
        constructor(public stringValue: string) {}
    };
    class UserName {
        constructor(public stringValue: string) {}
    }
 
    const validateUserName = (userName: UserName): true | ErrorMessage => {
        if (userName.stringValue.match(/^[a-z]+$/) === null) return `user name must be l/c: [${userName}]`;
        return true;
    }
    const validateAccountCode = (code: AccountCode): true | ErrorMessage => {
        if (code.stringValue.match(/^[AB][0-9]+$/) === null) return `account code not valid: [${code}]`;
        return true;
    }
    export const validateDetails = (userName: UserName, accountCode: AccountCode): void => {
        if (userName.stringValue === '') throw 'user name cannot be blank'; 
        if (accountCode.stringValue === '') throw 'account code cannot be blank'; 
        let msg = validateAccountCode(accountCode);
        if (msg !== true) throw msg;
        msg = validateUserName(accountCode);
        if (msg !== true) throw msg;
    }
    export const goodName: UserName = new UserName('bob');
    export const goodCode: AccountCode = new AccountCode('A1234');
    validateDetails(goodName, goodCode); // what bug makes this throw an exception ???
}